In [7]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [8]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [9]:
# import excel file StateGeneralbyPrecinctNov2016.xlsx
tn_2016=pd.read_excel('../data/2016_election/StateGeneralbyPrecinctNov2016.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','RNAME1','PARTY1','PVTALLY1',
                             'RNAME2','PARTY2','PVTALLY2','RNAME3','PARTY3','PVTALLY3','RNAME4','PARTY4','PVTALLY4',
                             'RNAME5','PARTY5','PVTALLY5','RNAME6','PARTY6','PVTALLY6','RNAME7','PARTY7','PVTALLY7',
                             'RNAME8','PARTY8','PVTALLY8','RNAME9','PARTY9','PVTALLY9','RNAME10','PARTY10','PVTALLY10'])
tn_2016

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9397    Wilson       Absentee   
9398    Wilson       Absentee   
9399    Wilson       Absentee   
9400    Wilson       Absentee   
9401    Wilson       Absentee   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9397                            United States President  2016-11-08   
9398                            United States President  2016-11-08   
9399  United States House of Representatives District 6  2016-11-08   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   

                              RNAME1       PARTY1  PVTALLY1  \
0                    Donald J. Trump   Republican      1222   
1     Write-In - Kyle Kenley Kopitke          NaN         0   
2                  Chuck Fleischmann   Republican      1124   
3                   Dennis H. Powers   Republican      1128   
4                    Donald J. Trump   Republican       243   
...                              ...          ...       ...   
9397                 Donald J. Trump   Republican       661   
9398  Write-In - Kyle Kenley Kopitke  Independent         0   
9399                     Diane Black   Republican       699   
9400                       Mark Pody   Republican       264   
9401                    Susan M Lynn   Republican       432   

                             RNAME2       PARTY2  PVTALLY2  \
0                   Hillary Clinton   Democratic     227.0   
1     Write-In - Laurence Kotlikoff          NaN       0.0   
2                    Melody Shekari   Democratic     195.0   
3                        Bob Fannon   Democratic     265.0   
4                   Hillary Clinton   Democratic      42.0   
...                             ...          ...       ...   
9397                Hillary Clinton   Democratic     397.0   
9398  Write-In - Laurence Kotlikoff  Independent       0.0   
9399                  David W. Kent   Democratic     297.0   
9400          Amelia Morrison Hipps   Democratic     142.0   
9401                  Trisha Farmer   Democratic     219.0   

                          RNAME3       PARTY3  PVTALLY3  \
0     "Rocky" Roque De La Fuente  Independent       2.0   
1      Write-In - David Limbaugh          NaN       0.0   
2                Topher Kersting  Independent      21.0   
3                            NaN          NaN       NaN   
4     "Rocky" Roque De La Fuente  Independent       3.0   
...                          ...          ...       ...   
9397  "Rocky" Roque De La Fuente  Independent       4.0   
9398   Write-In - David Limbaugh  Independent       0.0   
9399                  David Ross  Independent      71.0   
9400                         NaN          NaN       NaN   
9401                         NaN          NaN       NaN   

                        RNAME4       PARTY4  PVTALLY4  \
0                 Gary Johnson  Independent      32.0   
1     Write-In - Evan McMullin          NaN       3.0   
2           Cassandra Mitchell  Independent      36.0   
3                          NaN          NaN       NaN   
4                 Gary Johnson  Independent       5.0   
...                        ...          ...       ...   
9397              Gary Johnson  Independent      47.0   
9398  Write-In - Evan McMullin  Independent      14.0   
9399                       NaN    

In [10]:
tn_2016.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9402 entries, 0 to 9401
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      9402 non-null   object 
 1   PRECINCT    9402 non-null   object 
 2   OFFICENAME  9402 non-null   object 
 3   ELECTDATE   9402 non-null   object 
 4   RNAME1      9402 non-null   object 
 5   PARTY1      7561 non-null   object 
 6   PVTALLY1    9402 non-null   int64  
 7   RNAME2      7984 non-null   object 
 8   PARTY2      6088 non-null   object 
 9   PVTALLY2    7984 non-null   float64
 10  RNAME3      5738 non-null   object 
 11  PARTY3      3787 non-null   object 
 12  PVTALLY3    5738 non-null   float64
 13  RNAME4      4913 non-null   object 
 14  PARTY4      2726 non-null   object 
 15  PVTALLY4    4913 non-null   float64
 16  RNAME5      4567 non-null   object 
 17  PARTY5      2726 non-null   object 
 18  PVTALLY5    4567 non-null   float64
 19  RNAME6      2269 non-null  

In [11]:
# merge entries in all RNAME columns into new column "name"
tn_2016_namemelt=tn_2016.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['RNAME1','RNAME2','RNAME3','RNAME4','RNAME5','RNAME6','RNAME7','RNAME8','RNAME9','RNAME10'],value_name="name")
# drop column "variable" created by .melt
tn_2016_namemelt=tn_2016_namemelt.drop(['variable'],axis=1)                          
tn_2016_namemelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9400    Wilson       Absentee   
9401    Wilson       Absentee   
9402  Anderson  Andersonville   
9403  Anderson  Andersonville   
9404  Anderson  Andersonville   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   
9402                            United States President  2016-11-08   
9403                            United States President  2016-11-08   
9404  United States House of Representatives District 3  2016-11-08   

                                name  
0                    Donald J. Trump  
1     Write-In - Kyle Kenley Kopitke  
2                  Chuck Fleischmann  
3                   Dennis H. Powers  
4                    Donald J. Trump  
...                              ...  
9400                       Mark Pody  
9401                    Susan M Lynn  
9402                 Hillary Clinton  
9403   Write-In - Laurence Kotlikoff  
9404                  Melody Shekari  

[9405 rows x 5 columns]

In [12]:
# merge entries in all PARTY columns into new column "party"
tn_2016_partymelt=tn_2016.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PARTY1','PARTY2','PARTY3','PARTY4','PARTY5','PARTY6','PARTY7','PARTY8','PARTY9','PARTY10'],value_name="party")
# drop column "variable" created by .melt
tn_2016_partymelt=tn_2016_partymelt.drop(['variable'],axis=1)
tn_2016_partymelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9400    Wilson       Absentee   
9401    Wilson       Absentee   
9402  Anderson  Andersonville   
9403  Anderson  Andersonville   
9404  Anderson  Andersonville   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   
9402                            United States President  2016-11-08   
9403                            United States President  2016-11-08   
9404  United States House of Representatives District 3  2016-11-08   

           party  
0     Republican  
1            NaN  
2     Republican  
3     Republican  
4     Republican  
...          ...  
9400  Republican  
9401  Republican  
9402  Democratic  
9403         NaN  
9404  Democratic  

[9405 rows x 5 columns]

In [13]:
# merge entries in all PVTALLY columns into new column "vote_count"
tn_2016_pvtallymelt=tn_2016.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PVTALLY1','PVTALLY2','PVTALLY3','PVTALLY4','PVTALLY5','PVTALLY6','PVTALLY7','PVTALLY8','PVTALLY9','PVTALLY10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2016_pvtallymelt=tn_2016_pvtallymelt.drop(['variable'],axis=1)
tn_2016_pvtallymelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9400    Wilson       Absentee   
9401    Wilson       Absentee   
9402  Anderson  Andersonville   
9403  Anderson  Andersonville   
9404  Anderson  Andersonville   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   
9402                            United States President  2016-11-08   
9403                            United States President  2016-11-08   
9404  United States House of Representatives District 3  2016-11-08   

      vote_count  
0         1222.0  
1            0.0  
2         1124.0  
3         1128.0  
4          243.0  
...          ...  
9400       264.0  
9401       432.0  
9402       227.0  
9403         0.0  
9404       195.0  

[9405 rows x 5 columns]

In [14]:
# add "party" and "vote_count" columns to tn_2016_namemelt
party=tn_2016_partymelt['party']
pvtally=tn_2016_pvtallymelt['vote_count']
tn_2016_namemelt=tn_2016_namemelt.join(party).join(pvtally)
tn_2016_namemelt.head(3)

COUNTY       PRECINCT                                         OFFICENAME  \
0  Anderson  Andersonville                            United States President   
1  Anderson  Andersonville                            United States President   
2  Anderson  Andersonville  United States House of Representatives District 3   

    ELECTDATE                            name       party  vote_count  
0  2016-11-08                 Donald J. Trump  Republican      1222.0  
1  2016-11-08  Write-In - Kyle Kenley Kopitke         NaN         0.0  
2  2016-11-08               Chuck Fleischmann  Republican      1124.0

In [15]:
tn_2016_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94020 entries, 0 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      94020 non-null  object 
 1   PRECINCT    94020 non-null  object 
 2   OFFICENAME  94020 non-null  object 
 3   ELECTDATE   94020 non-null  object 
 4   name        43208 non-null  object 
 5   party       27969 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 5.0+ MB


In [16]:
# lower case column names
tn_2016_namemelt.columns=tn_2016_namemelt.columns.str.lower()
tn_2016_namemelt.head(2)

county       precinct               officename   electdate  \
0  Anderson  Andersonville  United States President  2016-11-08   
1  Anderson  Andersonville  United States President  2016-11-08   

                             name       party  vote_count  
0                 Donald J. Trump  Republican      1222.0  
1  Write-In - Kyle Kenley Kopitke         NaN         0.0

In [17]:
# change null values in "party" column to "no party"
tn_2016_namemelt['party'].fillna("No party", inplace = True)
tn_2016_namemelt.head(2)

county       precinct               officename   electdate  \
0  Anderson  Andersonville  United States President  2016-11-08   
1  Anderson  Andersonville  United States President  2016-11-08   

                             name       party  vote_count  
0                 Donald J. Trump  Republican      1222.0  
1  Write-In - Kyle Kenley Kopitke    No party         0.0

In [18]:
tn_2016_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94020 entries, 0 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      94020 non-null  object 
 1   precinct    94020 non-null  object 
 2   officename  94020 non-null  object 
 3   electdate   94020 non-null  object 
 4   name        43208 non-null  object 
 5   party       94020 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 5.0+ MB


In [19]:
# change null values in "vote_count" column to "0"
tn_2016_namemelt['vote_count'].fillna(0)

0        1222.0
1           0.0
2        1124.0
3        1128.0
4         243.0
          ...  
94015       0.0
94016       0.0
94017       0.0
94018       0.0
94019       0.0
Name: vote_count, Length: 94020, dtype: float64

In [20]:
tn_2016_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94020 entries, 0 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      94020 non-null  object 
 1   precinct    94020 non-null  object 
 2   officename  94020 non-null  object 
 3   electdate   94020 non-null  object 
 4   name        43208 non-null  object 
 5   party       94020 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 5.0+ MB


### 43208 non-null names and 43208 non-null vote counts
### Do the nulls in "name" correlate one to one with "no party"?

In [21]:
tn_2016_namemelt[tn_2016_namemelt['name'].isna()]

county             precinct  \
9573   Benton           1 Holladay   
9578   Benton     2 Camden Jr High   
9583   Benton   3 Camden City Hall   
9588   Benton  4 Camden Elementary   
9593   Benton            5 Library   
...       ...                  ...   
94014  Wilson          Provisional   
94016  Wilson             Absentee   
94017  Wilson             Absentee   
94018  Wilson             Absentee   
94019  Wilson             Absentee   

                                              officename   electdate name  \
9573                        Tennessee Senate District 24  2016-11-08  NaN   
9578                        Tennessee Senate District 24  2016-11-08  NaN   
9583                        Tennessee Senate District 24  2016-11-08  NaN   
9588                        Tennessee Senate District 24  2016-11-08  NaN   
9593                        Tennessee Senate District 24  2016-11-08  NaN   
...                                                  ...         ...  ...   
94014     Tennessee House of Representatives District 57  2016-11-08  NaN   
94016                            United States President  2016-11-08  NaN   
94017  United States House of Representatives District 6  2016-11-08  NaN   
94018     Tennessee House of Representatives District 46  2016-11-08  NaN   
94019     Tennessee House of Representatives District 57  2016-11-08  NaN   

          party  vote_count  
9573   No party         NaN  
9578   No party         NaN  
9583   No party         NaN  
9588   No party         NaN  
9593   No party         NaN  
...         ...         ...  
94014  No party         NaN  
94016  No party         NaN  
94017  No party         NaN  
94018  No party         NaN  
94019  No party         NaN  

[50812 rows x 7 columns]

In [22]:
tn_2016_namemelt[tn_2016_namemelt['name'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50812 entries, 9573 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      50812 non-null  object 
 1   precinct    50812 non-null  object 
 2   officename  50812 non-null  object 
 3   electdate   50812 non-null  object 
 4   name        0 non-null      object 
 5   party       50812 non-null  object 
 6   vote_count  0 non-null      float64
dtypes: float64(1), object(6)
memory usage: 3.1+ MB


### Looks safe to drop rows with no votes tallied

In [23]:
# drop rows with no name and vote count
tn_2016_final=tn_2016_namemelt.dropna()

In [24]:
tn_2016_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43208 entries, 0 to 94015
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      43208 non-null  object 
 1   precinct    43208 non-null  object 
 2   officename  43208 non-null  object 
 3   electdate   43208 non-null  object 
 4   name        43208 non-null  object 
 5   party       43208 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 2.6+ MB


In [25]:
# convert vote_count data from float64 to int 32 to to drop ".0"
tn_2016_final=tn_2016_final.astype({'vote_count':'int32'})
tn_2016_final

county       precinct  \
0      Anderson  Andersonville   
1      Anderson  Andersonville   
2      Anderson  Andersonville   
3      Anderson  Andersonville   
4      Anderson     Briceville   
...         ...            ...   
93997    Wilson           24-1   
94001    Wilson           25-1   
94005    Wilson           25-2   
94010    Wilson    Provisional   
94015    Wilson       Absentee   

                                              officename   electdate  \
0                                United States President  2016-11-08   
1                                United States President  2016-11-08   
2      United States House of Representatives District 3  2016-11-08   
3         Tennessee House of Representatives District 36  2016-11-08   
4                                United States President  2016-11-08   
...                                                  ...         ...   
93997                            United States President  2016-11-08   
94001                            United States President  2016-11-08   
94005                            United States President  2016-11-08   
94010                            United States President  2016-11-08   
94015                            United States President  2016-11-08   

                                 name        party  vote_count  
0                     Donald J. Trump   Republican        1222  
1      Write-In - Kyle Kenley Kopitke     No party           0  
2                   Chuck Fleischmann   Republican        1124  
3                    Dennis H. Powers   Republican        1128  
4                     Donald J. Trump   Republican         243  
...                               ...          ...         ...  
93997         Write-In - Tom Hoefling  Independent           0  
94001         Write-In - Tom Hoefling  Independent           0  
94005         Write-In - Tom Hoefling  Independent           0  
94010         Write-In - Tom Hoefling  Independent           0  
94015         Write-In - Tom Hoefling  Independent           0  

[43208 rows x 7 columns]

In [26]:
tn_2016_final.dtypes

county        object
precinct      object
officename    object
electdate     object
name          object
party         object
vote_count     int32
dtype: object

In [27]:
tn_2016_final['officename'].value_counts()

United States President                              30330
United States House of Representatives District 8     1729
United States House of Representatives District 3     1380
United States House of Representatives District 1      868
United States House of Representatives District 7      852
                                                     ...  
Tennessee House of Representatives District 99          12
Tennessee House of Representatives District 68          11
Tennessee House of Representatives District 1           11
Tennessee House of Representatives District 16          10
Tennessee House of Representatives District 24          10
Name: officename, Length: 125, dtype: int64

In [28]:
tn_2016_final['officename'].unique()

array(['United States President',
       'United States House of Representatives District 3',
       'Tennessee House of Representatives District 36',
       'Tennessee House of Representatives District 33',
       'United States House of Representatives District 4',
       'Tennessee Senate District 14',
       'Tennessee House of Representatives District 62',
       'United States House of Representatives District 7',
       'Tennessee Senate District 24',
       'Tennessee House of Representatives District 75',
       'United States House of Representatives District 8',
       'Tennessee House of Representatives District 31',
       'United States House of Representatives District 2',
       'Tennessee Senate District 2',
       'Tennessee House of Representatives District 8',
       'Tennessee House of Representatives District 20',
       'Tennessee Senate District 10',
       'Tennessee House of Representatives District 24',
       'Tennessee House of Representatives District 22',

In [29]:
tn_2016_final['vote_count'].value_counts()

0       16824
1        2367
2        1670
3        1193
4         930
        ...  
1742        1
1710        1
1614        1
1582        1
2353        1
Name: vote_count, Length: 2211, dtype: int64

### 16824 rows have a vote count of 0. Drop those rows

In [30]:
tn_2016_final=tn_2016_final[tn_2016_final['vote_count'] != 0]
tn_2016_final=tn_2016_final.reset_index(drop=True)
tn_2016_final.head()

county       precinct                                         officename  \
0  Anderson  Andersonville                            United States President   
1  Anderson  Andersonville  United States House of Representatives District 3   
2  Anderson  Andersonville     Tennessee House of Representatives District 36   
3  Anderson     Briceville                            United States President   
4  Anderson     Briceville  United States House of Representatives District 3   

    electdate               name       party  vote_count  
0  2016-11-08    Donald J. Trump  Republican        1222  
1  2016-11-08  Chuck Fleischmann  Republican        1124  
2  2016-11-08   Dennis H. Powers  Republican        1128  
3  2016-11-08    Donald J. Trump  Republican         243  
4  2016-11-08  Chuck Fleischmann  Republican         207

In [31]:
# confirm no 0's in vote count
tn_2016_final['vote_count'].value_counts()

1       2367
2       1670
3       1193
4        930
5        714
        ... 
1079       1
2998       1
2966       1
2934       1
2049       1
Name: vote_count, Length: 2210, dtype: int64

In [32]:
# create data frame tn_2016_pres for presidental election
tn_2016_pres_int=tn_2016_final[tn_2016_final['officename'].str.match('United States President')]
tn_2016_pres_int=tn_2016_pres_int.reset_index(drop=True)
tn_2016_pres_int

county       precinct               officename   electdate  \
0      Anderson  Andersonville  United States President  2016-11-08   
1      Anderson     Briceville  United States President  2016-11-08   
2      Anderson       Bull Run  United States President  2016-11-08   
3      Anderson        Clinton  United States President  2016-11-08   
4      Anderson   Clinton High  United States President  2016-11-08   
...         ...            ...                      ...         ...   
14085    Wilson            1-1  United States President  2016-11-08   
14086    Wilson           11-1  United States President  2016-11-08   
14087    Wilson           14-1  United States President  2016-11-08   
14088    Wilson           16-1  United States President  2016-11-08   
14089    Wilson           23-1  United States President  2016-11-08   

                          name        party  vote_count  
0              Donald J. Trump   Republican        1222  
1              Donald J. Trump   Republican         243  
2              Donald J. Trump   Republican        1397  
3              Donald J. Trump   Republican        1330  
4              Donald J. Trump   Republican         792  
...                        ...          ...         ...  
14085  Write-In - Tom Hoefling  Independent           1  
14086  Write-In - Tom Hoefling  Independent           2  
14087  Write-In - Tom Hoefling  Independent           1  
14088  Write-In - Tom Hoefling  Independent           1  
14089  Write-In - Tom Hoefling  Independent           1  

[14090 rows x 7 columns]

In [33]:
tn_2016_pres_int.head(20)

county        precinct               officename   electdate  \
0   Anderson   Andersonville  United States President  2016-11-08   
1   Anderson      Briceville  United States President  2016-11-08   
2   Anderson        Bull Run  United States President  2016-11-08   
3   Anderson         Clinton  United States President  2016-11-08   
4   Anderson    Clinton High  United States President  2016-11-08   
5   Anderson         Claxton  United States President  2016-11-08   
6   Anderson  Clinton Middle  United States President  2016-11-08   
7   Anderson    Dutch Valley  United States President  2016-11-08   
8   Anderson    Emory Valley  United States President  2016-11-08   
9   Anderson        Fairview  United States President  2016-11-08   
10  Anderson     Glen Alpine  United States President  2016-11-08   
11  Anderson        Glenwood  United States President  2016-11-08   
12  Anderson   Hendrix Creek  United States President  2016-11-08   
13  Anderson   Highland View  United States President  2016-11-08   
14  Anderson    Lake City MS  United States President  2016-11-08   
15  Anderson          Marlow  United States President  2016-11-08   
16  Anderson   North Clinton  United States President  2016-11-08   
17  Anderson          Norris  United States President  2016-11-08   
18  Anderson         Norwood  United States President  2016-11-08   
19  Anderson       Oak Ridge  United States President  2016-11-08   

               name       party  vote_count  
0   Donald J. Trump  Republican        1222  
1   Donald J. Trump  Republican         243  
2   Donald J. Trump  Republican        1397  
3   Donald J. Trump  Republican        1330  
4   Donald J. Trump  Republican         792  
5   Donald J. Trump  Republican        1319  
6   Donald J. Trump  Republican         453  
7   Donald J. Trump  Republican         271  
8   Donald J. Trump  Republican        1034  
9   Donald J. Trump  Republican         543  
10  Donald J. Trump  Republican        1149  
11  Donald J. Trump  Republican         763  
12  Donald J. Trump  Republican         849  
13  Donald J. Trump  Republican         299  
14  Donald J. Trump  Republican        1219  
15  Donald J. Trump  Republican         505  
16  Donald J. Trump  Republican         809  
17  Donald J. Trump  Republican         412  
18  Donald J. Trump  Republican         915  
19  Donald J. Trump  Republican         324

### Add column that has total vote count by count for use in calculating county vote percentages

In [34]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2016_pres_int=tn_2016_pres_int.groupby(['officename','county','name','party'], as_index=False)['vote_count'].sum()
tn_2016_pres=tn_2016_pres_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2016_pres.head(20)

officename      county                           name  \
0   United States President      Wilson                Donald J. Trump   
1   United States President      Wilson                Hillary Clinton   
2   United States President      Wilson                   Gary Johnson   
3   United States President      Wilson                     Jill Stein   
4   United States President      Wilson       Write-In - Evan McMullin   
5   United States President      Wilson                     Mike Smith   
6   United States President      Wilson                 Alyson Kennedy   
7   United States President      Wilson     "Rocky" Roque De La Fuente   
8   United States President      Wilson   Write-In - Darrell L. Castle   
9   United States President      Wilson        Write-In - Tom Hoefling   
10  United States President      Wilson  Write-In - Laurence Kotlikoff   
11  United States President  Williamson                Donald J. Trump   
12  United States President  Williamson                Hillary Clinton   
13  United States President  Williamson                   Gary Johnson   
14  United States President  Williamson       Write-In - Evan McMullin   
15  United States President  Williamson                     Jill Stein   
16  United States President  Williamson                     Mike Smith   
17  United States President  Williamson     "Rocky" Roque De La Fuente   
18  United States President  Williamson   Write-In - Darrell L. Castle   
19  United States President  Williamson                 Alyson Kennedy   

          party  vote_count  
0    Republican       39406  
1    Democratic       14385  
2   Independent        1807  
3   Independent         467  
4   Independent         280  
5   Independent         207  
6   Independent          83  
7   Independent          69  
8   Independent          23  
9   Independent           6  
10  Independent           1  
11   Republican       68212  
12   Democratic       31013  
13  Independent        4518  
14     No party        1371  
15  Independent         526  
16  Independent         318  
17  Independent         130  
18     No party         102  
19  Independent          67

In [35]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2016_pres.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0      Wilson       39406
1      Wilson       14385
2      Wilson        1807
3      Wilson         467
4      Wilson         280
..        ...         ...
884  Anderson         174
885  Anderson         115
886  Anderson          74
887  Anderson          42
888  Anderson          28

[889 rows x 2 columns]

In [36]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson       29881
Bedford        15355
Benton          6328
Bledsoe         4664
Blount         52208
               ...  
Wayne           5857
Weakley        12184
White           9825
Williamson    106271
Wilson         56734
Name: vote_count, Length: 95, dtype: int32

In [37]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2016_pres correlating with value in each row
tn_2016_pres['sum_county_vote']=tn_2016_pres['county'].map(county_vote_sum)
tn_2016_pres.head(20)

officename      county                           name  \
0   United States President      Wilson                Donald J. Trump   
1   United States President      Wilson                Hillary Clinton   
2   United States President      Wilson                   Gary Johnson   
3   United States President      Wilson                     Jill Stein   
4   United States President      Wilson       Write-In - Evan McMullin   
5   United States President      Wilson                     Mike Smith   
6   United States President      Wilson                 Alyson Kennedy   
7   United States President      Wilson     "Rocky" Roque De La Fuente   
8   United States President      Wilson   Write-In - Darrell L. Castle   
9   United States President      Wilson        Write-In - Tom Hoefling   
10  United States President      Wilson  Write-In - Laurence Kotlikoff   
11  United States President  Williamson                Donald J. Trump   
12  United States President  Williamson                Hillary Clinton   
13  United States President  Williamson                   Gary Johnson   
14  United States President  Williamson       Write-In - Evan McMullin   
15  United States President  Williamson                     Jill Stein   
16  United States President  Williamson                     Mike Smith   
17  United States President  Williamson     "Rocky" Roque De La Fuente   
18  United States President  Williamson   Write-In - Darrell L. Castle   
19  United States President  Williamson                 Alyson Kennedy   

          party  vote_count  sum_county_vote  
0    Republican       39406            56734  
1    Democratic       14385            56734  
2   Independent        1807            56734  
3   Independent         467            56734  
4   Independent         280            56734  
5   Independent         207            56734  
6   Independent          83            56734  
7   Independent          69            56734  
8   Independent          23            56734  
9   Independent           6            56734  
10  Independent           1            56734  
11   Republican       68212           106271  
12   Democratic       31013           106271  
13  Independent        4518           106271  
14     No party        1371           106271  
15  Independent         526           106271  
16  Independent         318           106271  
17  Independent         130           106271  
18     No party         102           106271  
19  Independent          67           106271

In [38]:
# create "pct_county_vote" column for tn_2016_pres
tn_2016_pres['pct_county_vote']=round(tn_2016_pres['vote_count']/tn_2016_pres['sum_county_vote']*100,2)
tn_2016_pres

officename    county                          name  \
0    United States President    Wilson               Donald J. Trump   
1    United States President    Wilson               Hillary Clinton   
2    United States President    Wilson                  Gary Johnson   
3    United States President    Wilson                    Jill Stein   
4    United States President    Wilson      Write-In - Evan McMullin   
..                       ...       ...                           ...   
884  United States President  Anderson                    Mike Smith   
885  United States President  Anderson      Write-In - Evan McMullin   
886  United States President  Anderson    "Rocky" Roque De La Fuente   
887  United States President  Anderson                Alyson Kennedy   
888  United States President  Anderson  Write-In - Darrell L. Castle   

           party  vote_count  sum_county_vote  pct_county_vote  
0     Republican       39406            56734            69.46  
1     Democratic       14385            56734            25.36  
2    Independent        1807            56734             3.19  
3    Independent         467            56734             0.82  
4    Independent         280            56734             0.49  
..           ...         ...              ...              ...  
884  Independent         174            29881             0.58  
885     No party         115            29881             0.38  
886  Independent          74            29881             0.25  
887  Independent          42            29881             0.14  
888     No party          28            29881             0.09  

[889 rows x 7 columns]

In [39]:
# how many parties?
tn_2016_pres['party'].value_counts()

Independent    504
No party       195
Republican      95
Democratic      95
Name: party, dtype: int64

### Create dataframe "tn_2016_county_pres_pct"  with percent values per party per county

In [40]:
# create new dataframe with reduced columns values only
tn_2016_pct_county=tn_2016_pres.loc[0:,['county','party','pct_county_vote']]
tn_2016_pct_county

county        party  pct_county_vote
0      Wilson   Republican            69.46
1      Wilson   Democratic            25.36
2      Wilson  Independent             3.19
3      Wilson  Independent             0.82
4      Wilson  Independent             0.49
..        ...          ...              ...
884  Anderson  Independent             0.58
885  Anderson     No party             0.38
886  Anderson  Independent             0.25
887  Anderson  Independent             0.14
888  Anderson     No party             0.09

[889 rows x 3 columns]

In [41]:
# create new dataframe with Republican values only
tn_2016_rep_pct=tn_2016_pct_county[tn_2016_pct_county['party']== "Republican"].reset_index(drop=True)
tn_2016_rep_pct

county       party  pct_county_vote
0       Wilson  Republican            69.46
1   Williamson  Republican            64.19
2        White  Republican            78.08
3      Weakley  Republican            73.93
4        Wayne  Republican            85.98
..         ...         ...              ...
90      Blount  Republican            71.72
91     Bledsoe  Republican            77.66
92      Benton  Republican            74.53
93     Bedford  Republican            74.80
94    Anderson  Republican            64.30

[95 rows x 3 columns]

In [42]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2016_rep_pct=tn_2016_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2016_rep_pct

county  Republican
0       Wilson       69.46
1   Williamson       64.19
2        White       78.08
3      Weakley       73.93
4        Wayne       85.98
..         ...         ...
90      Blount       71.72
91     Bledsoe       77.66
92      Benton       74.53
93     Bedford       74.80
94    Anderson       64.30

[95 rows x 2 columns]

In [43]:
# create new dataframe with Democratic values only
tn_2016_dem_pct=tn_2016_pct_county[tn_2016_pct_county['party']== "Democratic"].reset_index(drop=True)
tn_2016_dem_pct

county       party  pct_county_vote
0       Wilson  Democratic            25.36
1   Williamson  Democratic            29.18
2        White  Democratic            18.78
3      Weakley  Democratic            22.75
4        Wayne  Democratic            12.24
..         ...         ...              ...
90      Blount  Democratic            23.18
91     Bledsoe  Democratic            19.23
92      Benton  Democratic            23.29
93     Bedford  Democratic            22.11
94    Anderson  Democratic            30.16

[95 rows x 3 columns]

In [44]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2016_dem_pct=tn_2016_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2016_dem_pct

county  Democratic
0       Wilson       25.36
1   Williamson       29.18
2        White       18.78
3      Weakley       22.75
4        Wayne       12.24
..         ...         ...
90      Blount       23.18
91     Bledsoe       19.23
92      Benton       23.29
93     Bedford       22.11
94    Anderson       30.16

[95 rows x 2 columns]

In [45]:
# create new dataframe with non Republican or Democratic values only
tn_2016_other_pct=tn_2016_pct_county[(tn_2016_pct_county['party']=="Independent")|(tn_2016_pct_county['party']=="No party")].reset_index(drop=True)
tn_2016_other_pct

county        party  pct_county_vote
0      Wilson  Independent             3.19
1      Wilson  Independent             0.82
2      Wilson  Independent             0.49
3      Wilson  Independent             0.36
4      Wilson  Independent             0.15
..        ...          ...              ...
694  Anderson  Independent             0.58
695  Anderson     No party             0.38
696  Anderson  Independent             0.25
697  Anderson  Independent             0.14
698  Anderson     No party             0.09

[699 rows x 3 columns]

In [46]:
# sum by county
tn_2016_other_pct=tn_2016_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2016_other_pct

county  pct_county_vote
0     Anderson             5.53
1      Bedford             3.08
2       Benton             2.18
3      Bledsoe             3.11
4       Blount             5.10
..         ...              ...
90       Wayne             1.78
91     Weakley             3.32
92       White             3.14
93  Williamson             6.62
94      Wilson             5.18

[95 rows x 2 columns]

In [47]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2016_other_pct=tn_2016_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2016_other_pct

county  Other
0     Anderson   5.53
1      Bedford   3.08
2       Benton   2.18
3      Bledsoe   3.11
4       Blount   5.10
..         ...    ...
90       Wayne   1.78
91     Weakley   3.32
92       White   3.14
93  Williamson   6.62
94      Wilson   5.18

[95 rows x 2 columns]

In [101]:
# join tn_2016_rep, tn_2016_dem_pct and tn_2016_other_pct to form tn_2016_county_pres_pct
dem=tn_2016_dem_pct['Democratic']
other=tn_2016_other_pct['Other']
tn_2016_county_pres_pct=tn_2016_rep_pct.join(dem).join(other)
tn_2016_county_pres_pct

county  Republican  Democratic  Other
0       Wilson       69.46       25.36   5.53
1   Williamson       64.19       29.18   3.08
2        White       78.08       18.78   2.18
3      Weakley       73.93       22.75   3.11
4        Wayne       85.98       12.24   5.10
..         ...         ...         ...    ...
90      Blount       71.72       23.18   1.78
91     Bledsoe       77.66       19.23   3.32
92      Benton       74.53       23.29   3.14
93     Bedford       74.80       22.11   6.62
94    Anderson       64.30       30.16   5.18

[95 rows x 4 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [102]:
tn_2016_county_pres_pct['Rep_Margin']= tn_2016_county_pres_pct['Republican']-tn_2016_county_pres_pct['Democratic']
tn_2016_county_pres_pct.head(10)

county  Republican  Democratic  Other  Rep_Margin
0      Wilson       69.46       25.36   5.53       44.10
1  Williamson       64.19       29.18   3.08       35.01
2       White       78.08       18.78   2.18       59.30
3     Weakley       73.93       22.75   3.11       51.18
4       Wayne       85.98       12.24   5.10       73.74
5  Washington       68.85       26.18   4.72       42.67
6      Warren       70.09       25.97   3.30       44.12
7   Van Buren       75.24       22.28   3.39       52.96
8       Union       80.89       16.20   2.89       64.69
9      Unicoi       78.82       17.54   3.47       61.28

In [104]:
tn_2016_county_pres_pct.to_excel(r'../data/2016_election/tn_2016_county_pres_pct.xlsx',index=False)